### ISSUE TITLE
Using batchQueries to solve the n+1 problem (mongo)

### ISSUE THREAD

**Issue by ruudvanbuul** [doug-martin/nestjs-query#1352]
I was doing some testing of quering by reference and noticed that there is no solution for the n+1 problem in mongoose/typegoose. I saw this issue (https://github.com/doug-martin/nestjs-query/issues/356) which explains the exact same problem but in typeorm, this has been fixed in the meantime with the batchQueryRelations function (https://github.com/doug-martin/nestjs-query/blob/fb6dc5455a412f16acc10c436881bf6fd139b34e/packages/query-sequelize/src/services/relation-query.service.ts#L343) in relation-query.service.ts of the type-orm package.

I would be nice to have something similar in mongo, otherwise queries with references will get expensive quickly.

**Comment by ruudvanbuul** [comment]
For the time being the documentation at https://doug-martin.github.io/nestjs-query/docs/introduction/getting-started should probably also state: "Built in dataloader avoiding the n+1 problem (n+1 is only fixen in typeorm atm)" instead of "Built in dataloader avoiding the n+1 problem"

**Comment by doug-martin** [comment]
@ruudvanbuul good point on calling out the batch optimization with typeorm, the other adapters are a lot harder to do this with. I'll add a note to the documentation.

**Comment by ruudvanbuul** [comment]
@doug-martin What makes it so much harder in Mongo? I might try and fabricate a PR if we need this functionality in mongo

**Comment by smolinari** [comment]
My 2 cents.

Aside from the problem of resolving refs, trying to normalize data like in a relational database isn't always the wise choice in No-SQL and often denormalizing is better for query optimization. For example, the given structure of accommodations to features to categories is, I believe, one level too deep. Categories can be stored in the Features for sure. Storing a proper category or categories within a feature can be controlled on the insertion side of features. So, you'd still need a Categories collection to bump up the entries against. There just wouldn't be any references. And, if you did need references to categories, you'd probably want a many-to-many junction collection for that, because many features (or other entities) could fall under multiple categories. Or, you'd need an array of category references within each entity. There are multiple ways to Rome for this, which doesn't help make the solution straightforward. 

All that said, I believe, either population or aggregation (with `$unionWith`) could be the answer to getting referenced documents resolved. They handle both the n+1 issue albeit in different ways. I believe they both can handle ordering and limiting. The only thing I'd need to know/ need to test is if `$unionWith` can be used with the same collection. I haven't used it myself, so not sure and the examples I've seen all union on different collections. 

So, tl;dr; it should be possible to resolve refs, I'm just not sure of the possible solution and it working for all necessary data modeling scenarios. It might require specific modeling, which might be too restrictive. It definitely needs some extra contemplation.  

The other question I'd have is where could or can even DataLoader fit in? Instead of trying solve the data merging within the database, couldn't it be done on the resolver side and let GraphQL do the merging?

Scott
